File: leetcode/0001-1000/0101-0200/0199-binary-tree-right-side-view/BinaryTreeRightSideView.java
import java.util.ArrayList;
import java.util.List;

public class BinaryTreeRightSideView {
  public static void main(String[] args) {
    TreeNode root = new TreeNode(
        1,
        new TreeNode(
            2,
            new TreeNode(
                4,
                new TreeNode(5),
                null),
            null),
        new TreeNode(3));

    System.out.println(new Solution().rightSideView(root));
  }
}

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {
  }

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  void bfs(TreeNode node, int depth, List<Integer> rights) {
    if (node == null) {
      return;
    }

    if (depth == rights.size()) {
      rights.add(node.val);
    }

    if (node.right != null) {
      bfs(node.right, depth + 1, rights);
    }

    if (node.left != null) {
      bfs(node.left, depth + 1, rights);
    }
  }

  public List<Integer> rightSideView(TreeNode root) {
    List<Integer> rights = new ArrayList<>();

    bfs(root, 0, rights);

    return rights;
  }
}

File: leetcode/0001-1000/0101-0200/0199-binary-tree-right-side-view/BinaryTreeRightSideView.ts
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

function rightSideView(root: TreeNode | null): number[] {
  if (root == null) {
    return [];
  }

  const answer = {};

  const queue: { node: TreeNode, depth: number }[] = [{ node: root, depth: 0 }];

  while (queue.length > 0) {
    const { node, depth } = queue.shift()!;

    answer[depth] = node.val;

    if (node.left) {
      queue.push({ node: node.left, depth: depth + 1 });
    }

    if (node.right) {
      queue.push({ node: node.right, depth: depth + 1 });
    }
  }

  return Object.values(answer);
};

const root = new TreeNode(
  1,
  new TreeNode(
    2,
    new TreeNode(
      4,
      new TreeNode(5))
  ),
  new TreeNode(3)
);

console.log(rightSideView(root));
